File: emaar.js
const request = require('request');
const sql = require("mssql/msnodesqlv8");
const { json } = require('express');


// ----!!!!BU KODU ÇALIŞTIRMAK İÇİN GÜN AYARINI DEĞİŞTİRMENİZ GEREKMEKTEDİR!!!!--- //


function roundUpToDecimal(num, decimalPlaces) {
    const factor = Math.pow(10, decimalPlaces);
    return Math.ceil(num * factor) / factor;
}

// Today's Date
const today = new Date();
const mmToday = String(today.getMonth() + 1).padStart(2, '0'); // Get month, add 1 since months are 0-based
const ddToday = String(today.getDate()).padStart(2, '0'); // Get day of the month
const yyyyToday = today.getFullYear(); // Get year

const formattedToday = `${mmToday}/${ddToday}/${yyyyToday}`;


// Yesterday's Date
const yesterday = new Date(today);
yesterday.setDate(today.getDate() - 1); // Subtract one day

const mmYesterday = String(yesterday.getMonth() + 1).padStart(2, '0');
const ddYesterday = String(yesterday.getDate()).padStart(2, '0');
const yyyyYesterday = yesterday.getFullYear();

const formattedYesterday = `${mmYesterday}/${ddYesterday}/${yyyyYesterday}`;


const config = {
    server: "localhost",
    database: "NCR",
    driver: "msnodesqlv8",
    options: {
        trustedConnection: true // Uses,                                                                                                             
    }
};

async function connectDB() {
    try {
        let pool = await sql.connect(config);
        console.log("Connected to MSSQL using Windows Authentication!");

        const toplamFis = await pool.request().query('SELECT COUNT(DT) AS Fis_Toplam FROM NCR.dbo.TDT_HEADER WHERE STR = 1033 AND AMT > 0 AND CAST(DT AS DATE) = CAST(DATEADD(DAY, -1, GETDATE()) AS DATE);');
        const gunlukCiro = await pool.request().query('SELECT SUM(AMT) AS Satis_Toplam FROM NCR.dbo.TDT_HEADER WHERE STR = 1033 AND AMT != 0 AND CAST(DT AS DATE) = CAST(DATEADD(DAY, -1, GETDATE()) AS DATE)');
        const toplamKDV = await pool.request().query('SELECT SUM(T_VAT) AS KDV_Toplam FROM NCR.dbo.TDT_TRX WHERE T_STR = 1033 AND T_AMOUNT != 0 AND CAST(DT AS DATE) = CAST(DATEADD(DAY, -1, GETDATE()) AS DATE)')

        gunlukCiro = roundUpToDecimal(gunlukCiro.recordset[0].Satis_Toplam - toplamKDV.recordset[0].KDV_Toplam, 2).toString()
        toplamFis = toplamFis.recordset[0].Fis_Toplam.toString()

        console.log({
            "PropertyCode": "ESM",
            "LeaseName": "NEZİH",
            "LeaseCode": "t0000042",
            "SaleType": "otherme",
            "SalesFromDate": formattedYesterday,
            "NetSalesAmount": gunlukCiro,
            "NoofTransactions": toplamFis,
            "SalesFrequency": "Daily"
        })
        sql.close();

    } catch (err) {
        console.error("Database connection failed!", err);
    }
}


function execute() {
    const options = {
        url: "https://api.emaar.com/emaar/trky/sales",
        method: "POST",
        json: [
            {
        "PropertyCode": "ESM",
        "LeaseName": "NEZİH",
        "LeaseCode": "t0000042",
        "SaleType": "otherme",
        "SalesFromDate": formattedYesterday,
        "NetSalesAmount": gunlukCiro,
        "NoofTransactions": toplamFis,
        "SalesFrequency": "Daily"
    }
    ],
    headers: {
        "x-api-key": "JfSxyPg2Qp",
      "Content-Type": "application/json"
    }
  };

  request(options, function (err, res, body) {
      if (err) {
          console.error("Error:", err);
        } else {
            console.log("Response:", body);
        }
    });
}

connectDB();
// execute();

File: Emaar_Final/emaar.py
import pyodbc
import requests
import datetime

# Dünün tarihini hesapla
yesterday = datetime.date.today() - datetime.timedelta(days=1)
formatted_yesterday = yesterday.strftime('%m/%d/%Y')

# MSSQL bağlantı bilgileri
server = '10.0.0.64'
database = 'NCR'
conn_str = (
    f'DRIVER={{SQL Server}};'
    f'SERVER={server};'
    f'DATABASE={database};'
    f'Trusted_Connection=yes;'
)

# Veritabanına bağlan ve sorguları çalıştır
try:
    conn = pyodbc.connect(conn_str)
    cursor = conn.cursor()

    cursor.execute("""
        SELECT COUNT(DT) AS Fis_Toplam
        FROM NCR.dbo.TDT_HEADER
        WHERE STR = 1033 AND AMT > 0 
        AND CAST(DT AS DATE) = CAST(DATEADD(DAY, -1, GETDATE()) AS DATE);
    """)
    toplam_fis = cursor.fetchone()[0]

    cursor.execute("""
        SELECT SUM(AMT) AS Satis_Toplam
        FROM NCR.dbo.TDT_HEADER
        WHERE STR = 1033 AND AMT != 0 
        AND CAST(DT AS DATE) = CAST(DATEADD(DAY, -1, GETDATE()) AS DATE);
    """)
    gunluk_ciro = cursor.fetchone()[0]

    cursor.execute("""
        SELECT SUM(T_VAT) AS KDV_Toplam
        FROM NCR.dbo.TDT_TRX
        WHERE T_STR = 1033 AND T_AMOUNT != 0 
        AND CAST(DT AS DATE) = CAST(DATEADD(DAY, -1, GETDATE()) AS DATE);
    """)
    toplam_kdv = cursor.fetchone()[0]
    
    # Net Satış Miktarını hesapla
    net_satis = round(gunluk_ciro - toplam_kdv, 2)

    # Bağlantıyı kapat
    conn.close()
    
    # JSON verisini oluştur
    data = [{
        "PropertyCode": "ESM",
        "LeaseName": "NEZİH",
        "LeaseCode": "t0000042",
        "SaleType": "otherme",
        "SalesFromDate": formatted_yesterday,
        "NetSalesAmount": str(net_satis),
        "NoofTransactions": str(toplam_fis),
        "SalesFrequency": "Daily"
    }]

    # Çıktıyı ekrana yazdır
    print("Gönderilecek Veri:", data)

    # API'ye POST isteği gönder
    headers = {
        "x-api-key": "JfSxyPg2Qp",
        "Content-Type": "application/json"
    }

    response = requests.post("https://api.emaar.com/emaar/trky/sales", json=data, headers=headers)
    print("API Response:", response.text)

except Exception as e:
    print("Hata oluştu:", e)


